### PR TITLE
make design changes and add summary data

### PR DESCRIPTION
**Pull request by hanakamer**
## Summary

this code is to redesign the home page.

## Learnings

giving jsx chunks as props to parents in order to make the layout was a small wow moment:)

## Struggles

I struggled with the layout of the main container. I wasnt able to give bottom margin to the body.

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,15 +1,15 @@
 import React from 'react';
-import Styles from './App.module.css';
 import { CreateDay } from './pages/CreateDay';
 import { EditDay } from './pages/EditDay';
 import { Home } from './pages/Home';
 import { BrowserRouter as Router, Routes, Route } from 'react-router-dom';
 import { SavedGridProvider } from './contexts/SavedGridContext';
+import generalStyle from './common/general.module.css';
```
Comment (sandrina-p) [inline src/App.js]: chore: `StylesGeneral` is a better name because, similar to all the others, it starts with `Styles`, and it triggers the autocomplete in your editor while typing the word :)

File: src/common/general.module.css [hunk 2]
```diff
@@ -1,15 +1,40 @@
...
 }
-.sectionContainer {
+.verticalAlignMediumContainer {
+  display: grid;
+  justify-content: center;
+  align-content: center;
+  gap: 10px;
+  grid-auto-flow: row;
+}
+.verticalAlignSmallContainer {
```
Comment (sandrina-p) [inline src/common/general.module.css]: todo: Have you learned about CSS methology called BEM? You can use it here with this 3 classes. You can also use CSS variables to replace `gap` value in each variant. Give it a try here, pls :)

File: src/components/Day/DayView.js [hunk 3]
```diff
@@ -1,21 +1,20 @@
 import React from 'react';
 import PropTypes from 'prop-types';
-import { Cell } from '../Cell';
+import { SummaryCell } from '../SummaryCell';
```
Comment (sandrina-p) [inline src/components/Day/DayView.js:3]: bug: I cannot run the project on my side. Have you committed this file?

image

File: src/pages/Home/Home.js [hunk 4]
```diff
@@ -10,13 +10,10 @@ function Home() {
 
   return (
     <>
-      <p>{savedGrid && <span>{savedGrid.length}</span>}</p>
       <Days days={savedGrid} />
```
Comment (sandrina-p) [inline src/pages/Home/Home.js:14]: todo: It's Missing empty state "No sleep recorded yet"

File: src/components/Days/Days.js [hunk 5]
```diff
@@ -17,16 +20,30 @@ function Days({ days = [], ...props }) {
...
+                summary={Object.keys(summary).map((mode, i) => {
+                  return (
+                    <div key={i}>
+                      {SleepModes[mode]}: {summary[mode]}
+                    </div>
+                  );
+                })}
+                buttons={
+                  <>
+                    <Button name={'delete'} onClick={() => handleDelete(date)} />
```
Comment (sandrina-p) [inline src/components/Days/Days.js]: issue: The buttons here are reversed. It should be [Edit on top](https://user-images.githubusercontent.com/14869087/159657822-c26cbbc6-9989-4ce3-a267-7af03d305200.png)

File: src/pages/Home/Home.js [hunk 6]
```diff
@@ -10,13 +10,10 @@ function Home() {
...
-      {
-        <div className={generalStyles.sectionContainer}>
-          <Link to="/createDay"> Record Sleep</Link>
-        </div>
-      }
+      <Link to="/createDay" className={generalStyles.link}>
+        Record Sleep
```
Comment (sandrina-p) [inline src/pages/Home/Home.js:16]: todo: The button should be bold

File: src/components/Days/Days.js [hunk 7]
```diff
@@ -17,16 +20,30 @@ function Days({ days = [], ...props }) {
...
-                <Button name={'edit'} onClick={() => handleEdit(date)} />
-              </React.Fragment>
+              <Card
+                key={date}
+                data={<DayView data={data} deleteDay={() => handleDelete(date)}></DayView>}
+                date={date}
+                summary={Object.keys(summary).map((mode, i) => {
+                  return (
+                    <div key={i}>
+                      {SleepModes[mode]}: {summary[mode]}
```
Comment (sandrina-p) [inline src/components/Days/Days.js]: issue: The amount should be bold

File: src/components/EditScreen/EditScreen.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import PropTypes from 'prop-types';
```
Comment (sandrina-p) [inline src/components/EditScreen/EditScreen.js:1]: question: What's this component for? It seems leftovers

File: src/contexts/SavedGridContext.js [hunk 9]
```diff
@@ -13,9 +15,21 @@ export function SavedGridProvider({ children }) {
 
   function saveGrid(newDayData, newDate) {
     setSavedGrid((prev) => {
+      const summary = calculateSummaryOfSleep(newDayData);
+      console.log(summary);
+      Object.keys(summary).map((key) => {
```
Comment (sandrina-p) [inline src/contexts/SavedGridContext.js:20]: question: Why did you decide to create the summary on state?

Do you really need it here? Have you considered instead, to "create" the summary on runtime inside the `Days.jsx` component?

Advantages:
- Network: The data stored on localStorage (or sent to the server) is smaller, which makes it faster
- Separation of responsibilities: You can create multiple "representations" of the data (summary, analytics, etc...) in the component that need it, no need to be in the Local Storage.

File: src/utils/utils.js [hunk 10]
```diff
@@ -1,16 +1,27 @@
...
-export function minsToHours(totalminutes) {
+export function minsToTime(totalminutes) {
   const hours = Math.floor(totalminutes / 60);
   const minutes = totalminutes % 60;
   let resultH = hours > 23 ? hours - 24 : hours;
   resultH = resultH.toString().length < 2 ? `0${resultH}` : resultH;
   let resultM = minutes.toString().length < 2 ? `0${minutes}` : minutes;
   return `${resultH}:${resultM}`;
 }
+export function minsToHours(totalminutes) {
```
Comment (sandrina-p) [inline src/utils/utils.js]: chore: `totalMinutes` in camelCase to follow the same name convention as in the other variables :)

File: src/utils/utils.js [hunk 11]
```diff
@@ -26,3 +37,20 @@ export function createNewDayCells() {
...
+      map[val] = 1;
+    } else {
+      ++map[val];
+    }
+    return map;
+  };
+  const result = cells.map((cell) => cell.mode).reduce(reducer, {});
+  return result;
+}
+export const SleepModes = {
```
Comment (sandrina-p) [inline src/utils/utils.js]: Well done here with this const of sleep modes!

suggestion: Usually, it's common to make this type of variable in uppercase `SLEEP_MODES` to be clear that it's a "enum" of fixed variables used across the app.

File: src/pages/CreateDay/CreateDay.js [hunk 12]
```diff
@@ -55,33 +55,28 @@ function CreateDay() {
     setStartDate(newDate);
   }
   return (
-    <>
-      <div className={GeneralStyles.sectionContainer}>
-        <Day row={grid} mode={mode} changeRow={setGrid} />
-      </div>
-      {isPresent && (
-        <div className={GeneralStyles.sectionContainer}>
+    <EditScreen
```
Comment (sandrina-p) [inline src/pages/CreateDay/CreateDay.js:58]: question: What's the advantage, in this specific case of creating `EditScreen` component? It's used in only one place, and creates an extra level of _prop drilling_. What were the advantages you saw here?

File: src/components/FieldSleepMode/FieldSleepMode.js [hunk 13]
```diff
@@ -1,20 +1,21 @@
 import React from 'react';
 import { RadioButton } from '../RadioButton';
 import PropTypes from 'prop-types';
+import GeneralStyles from '../../common/general.module.css';
 
 function FieldSleepMode({ onChange }) {
   return (
-    <React.Fragment>
+    <div className={GeneralStyles.horizontalAlignContainer}>
       <RadioButton onChange={onChange} value="active">
```
Comment (sandrina-p) [inline src/components/FieldSleepMode/FieldSleepMode.js:9]: todo: The style of the radio inputs is incorrect.

[solution](https://user-images.githubusercontent.com/14869087/159661259-eba66c77-3d70-4b87-b617-a0fdeb2e74a5.png) vs [design](https://user-images.githubusercontent.com/14869087/159661315-179565df-bcce-4584-91c5-278ab4fd072b.png)

File: src/components/Card/Card.module.css [hunk 14]
```diff
@@ -0,0 +1,12 @@
+.card {
+  border: solid black 1px;
+  border-radius: 25px;
```
Comment (sandrina-p) [inline src/components/Card/Card.module.css]: question: The bigger border-radius compared to the design, was it on purpose?

File: src/components/Card/Card.module.css [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+  padding: 10px;
+  display: grid;
+  grid-template-columns: 70px 160px 100px;
```
Comment (sandrina-p) [inline src/components/Card/Card.module.css]: style: In web, especially with responsiveness, I tend to avoid fixed units for box sizes, as much as possible. Why did you need this here? The size of the elements for each column + size `auto` + `align-content: space-evenly;` + some `gap` would be enough, no? :)

File: src/components/Button/Button.module.css [hunk 16]
```diff
@@ -4,8 +4,8 @@
   font-weight: 800;
   min-height: 35px;
   min-width: 40px;
-  border: 1px solid #3f3feb;
-  color: #3f3feb;
+  border: 1px solid #000;
+  color: #000;
```
Comment (sandrina-p) [inline src/components/Button/Button.module.css]: todo: I recommend you to start using CSS variables as a way to group/reuse all the colors used in this app :)

File: src/common/general.module.css [hunk 17]
```diff
@@ -23,3 +48,13 @@
   width: 1px !important;
   white-space: nowrap !important;
 }
+.link {
```
Comment (sandrina-p) [inline src/common/general.module.css]: question: Is this class being used? I didn't find it :/

File: src/components/EditScreen/EditScreen.js (hunk 8 above)
Comment (hanakamer) [reply src/components/EditScreen/EditScreen.js:1]: this is inorder to style both editday and createady using 1 component

File: src/pages/CreateDay/CreateDay.js (hunk 12 above)
Comment (hanakamer) [reply src/pages/CreateDay/CreateDay.js:58]: to make the styling of createday and editday same

File: src/common/general.module.css [hunk 18]
```diff
@@ -1,40 +1,11 @@
 .container {
-  display: grid;
-  place-items: center;
-  align-content: center;
-  margin: 50px;
-  gap: 10px;
+  display: flex;
+  flex-direction: column;
+  align-items: center;
+  padding: 5% 3%;
```
Comment (sandrina-p) [inline src/common/general.module.css:5]: question: I'm curious, why did you choose % value here?

File: src/components/RadioButton/RadioButton.module.css [hunk 19]
```diff
@@ -1,62 +1,62 @@
...
+  height: 12px;
+  border-radius: 15px;
+  top: -2px;
+  left: -2px;
+  position: relative;
+  background-color: var(--color-white);
+  content: '';
+  display: inline-block;
+  visibility: visible;
+  border: 3px solid;
```
Comment (sandrina-p) [inline src/components/RadioButton/RadioButton.module.css:15]: issue: The cause of the problem of the input not being centered, as you mentioned in the video is... `box-sizing`! Have you read the article I shared about [box-sizing](https://css-tricks.com/box-sizing/)? You'll notice a part talking about pseudo selectors. You'll notice that these elements by default don't have the correct box-sizing, which makes math harder.

File: src/common/general.module.css (hunk 18 above)
Comment (hanakamer) [reply src/common/general.module.css:5]: My aim was to make the padding more responsive 🤔

File: src/components/RadioButton/RadioButton.module.css [hunk 20]
```diff
@@ -1,62 +1,64 @@
...
+  border-radius: 50%;
+  top: 50%;
+  left: -15px;
+  transform: translate(-50%, -50%);
+  position: absolute;
+  display: inline-block;
+  content: '';
+  visibility: visible;
+}
+.radioButton #active + label:before {
```
Comment (sandrina-p) [inline src/components/RadioButton/RadioButton.module.css:32]: Well done with the pseudoclasses here!

Now the question is: Can you find a way to simplify this without having to repeat the same styles multiple times? Hint: Using CSS variables 💡

File: src/components/Day/index.test.js [hunk 21]
```diff
@@ -0,0 +1,23 @@
...
+    const result = sum(3, 4);
+    const outcome = 7;
+
+    expect(result).toBe(outcome);
+  });
+
+  it('should return errror, given negative numbers', () => {
+    expect(sum(-3, 4)).toBe('ERROR!');
+  });
+});
```
Comment (sandrina-p) [inline src/components/Day/index.test.js]: This was just an example, no need to push this file, you can delete it :)

File: src/utils/utils.test.js [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    mode: 'active'
+  },
+  {
+    startIndex: 8,
+    endIndex: 10,
+    mode: 'sleep'
+  }
+];
+describe('createCells()', () => {
+  it('works', () => {
```
Comment (sandrina-p) [inline src/utils/utils.test.js]: IRL just saying `'works'` is not very intuitive. Tests are meant to describe what the code does. For example: `returns cells grouped by mode`. Same for the other tests.

File: src/components/SummaryCell/SummaryCell.module.css [hunk 23]
```diff
@@ -0,0 +1,16 @@
+.cell {
+  width: 25px;
+  height: 1px;
+  opacity: 1;
+  border-radius: 3px;
+  margin: 1px;
```
Comment (sandrina-p) [inline src/components/SummaryCell/SummaryCell.module.css]: question: Why did you prefer using `margin: 1px` instead of `gap: 2px` at the parent?

File: src/utils/utils.js [hunk 24]
```diff
@@ -26,3 +37,59 @@ export function createNewDayCells() {
...
+      ++map[val];
+    }
+    return map;
+  };
+  const result = cells.map((cell) => cell.mode).reduce(reducer, {});
+  return result;
+}
+export function clusterCells(cells) {
+  let currentMode = null;
+  let obj = {
```
Comment (sandrina-p) [inline src/utils/utils.js:54]: suggestion: Aim to give good names as much as possible. Here could be `clusterBase`

File: src/utils/utils.js [hunk 25]
```diff
@@ -26,3 +37,59 @@ export function createNewDayCells() {
...
+  const reducer = (map, val) => {
+    if (map[val] == null) {
+      map[val] = 1;
+    } else {
+      ++map[val];
+    }
+    return map;
+  };
+  const result = cells.map((cell) => cell.mode).reduce(reducer, {});
+  return result;
```
Comment (sandrina-p) [inline src/utils/utils.js:50]: This util could also be tested to guarantee it's working well.

File: src/utils/utils.js [hunk 26]
```diff
@@ -26,3 +37,59 @@ export function createNewDayCells() {
...
+    }
+    if (currentMode === null) {
+      currentMode = current.mode;
+      obj = { ...obj, startIndex: i, mode: current.mode };
+    }
+
+    return result;
+  }, []);
+
+  return clusteredCells;
```
Comment (sandrina-p) [inline src/utils/utils.js:83]: bug: Take a look at [this scenario](https://user-images.githubusercontent.com/14869087/164108631-eff68bc9-5af3-4487-94b7-0cec14f05abc.png) — you'll notice some gaps between blockers that are bigger than expected. You can reproduce it by recording an "active" block of just 15min.

Do you understand why? How would you solve it?

File: src/utils/utils.js [hunk 27]
```diff
@@ -26,3 +37,59 @@ export function createNewDayCells() {
...
+}
+export function calculatePercentage(partialVal, totalVal) {
```
Comment (sandrina-p) [inline src/utils/utils.js:85]: question: Are you familiar with [JSDoc](https://jsdoc.app/)? It's a common tool to document javascript functions, especially helpers like in this file. The most common is to document description and arguments. But you can also document `return` and give examples. Like this:

```js
/**
 * Returns percentage of a val based on total
 * @param {Number} partialVal - initial value
 * @param {Number} totalVal - total value
 * @return {Number}
 * @example (1, 2) -> 50
 * @example (2, 8) -> 25
 */
export function calculatePercentage(partialVal, totalVal) {
  // ....
}

```

File: src/utils/utils.js (hunk 26 above)
Comment (sandrina-p) [reply src/utils/utils.js:83]: Perhaps we should show only 3 clusters, summing the total of each mode... But I let you decide, which one do you prefer the most?
